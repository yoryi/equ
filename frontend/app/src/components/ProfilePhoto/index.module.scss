@import "src/styles/mixins/breakpoints";

:local(.container) {
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 64px;
  min-width: 64px;
  height: 64px;
  border-radius: 50%;
  box-shadow: 0 2px 4px rgba(31, 36, 43, 0.12), 0 0 8px rgba(31, 36, 43, 0.08);

  @include breakpoint('small') {
    width: 48px;
    min-width: 48px;
    height: 48px;
  }

  img {
    width: 100%;
    height: 100%;
    max-width: 64px;
    max-height: 64px;
    border-radius: 50%;

    &:local(.border) {
      position: absolute;
      top: 50%;
      left: 50%;
      transform: translate(-50%, -50%);
      max-width: 60px;
      max-height: 60px;
    }
  }

  :local(.initials) {
    font-family: Telegraf, sans-serif;
    font-size: 24px;
    line-height: 24px;
    font-weight: bold;
    color: black;

    @include breakpoint('small') {
      font-size: 18px;
      line-height: 18px;
    }
  }

  :local(.ring) {
    position: absolute;
    z-index: -1;

    path {
      shape-rendering: geometricPrecision;
    }
  }
}
