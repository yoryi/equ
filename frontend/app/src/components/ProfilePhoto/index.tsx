import * as mui from "@material-ui/core"
import classNames from "classnames"
import * as React from "react"
import { useSelector } from "react-redux"

import { INTERESTS_COLOR } from "../../const/interests"
import {
  DreamSchoolNotification,
  FollowNotification,
  FollowRequestNotification,
  Profile,
  ReduxState,
  Reference,
  Role,
  StudentNotificationSender,
  StudentProfile,
  UnfollowRequestNotification,
  University,
  UniversityProfile,
} from "../../store/types"
import getAvatar from "../../utils/getAvatar"
import { nameAbbreviation } from "../../utils/nameAbbreviation"
import Styles from "./index.module.scss"

interface ProfilePhotoProps {
  profile: Profile
  student?: undefined
  university?: undefined
  notification?: undefined
  reference?: undefined
  loading?: undefined
}

interface StudentProfilePhotoProps {
  profile?: undefined
  student: StudentProfile
  university?: undefined
  notification?: undefined
  reference?: undefined
  loading?: undefined
}

interface UniversityProfilePhotoProps {
  profile?: undefined
  student?: undefined
  university: University | UniversityProfile
  notification?: undefined
  reference?: undefined
  loading?: undefined
}

interface NotificationProfilePhotoProps {
  profile?: undefined
  student?: undefined
  university?: undefined
  notification:
    | FollowRequestNotification
    | FollowNotification
    | UnfollowRequestNotification
    | DreamSchoolNotification
  reference?: undefined
  loading?: undefined
}

interface ReferenceProfilePhotoProps {
  profile?: undefined
  student?: undefined
  university?: undefined
  notification?: undefined
  reference: Reference
  loading?: undefined
}

interface LoadingProfilePhotoProps {
  profile?: undefined
  student?: undefined
  university?: undefined
  notification?: undefined
  reference?: undefined
  loading: true
}

const ProfilePhoto: React.FC<
  | ProfilePhotoProps
  | StudentProfilePhotoProps
  | UniversityProfilePhotoProps
  | NotificationProfilePhotoProps
  | ReferenceProfilePhotoProps
  | LoadingProfilePhotoProps
> = ({
  profile,
  student,
  university,
  notification,
  reference,
  loading = false,
}) => {
  const mediaToken = useSelector<ReduxState, string | null>(
    (state) => state.media.mediaToken,
  )

  let avatar: string | null
  if (
    (notification?.sender.role === Role.Student &&
      notification.sender.extraData.avatar) ||
    profile?.avatar?.path ||
    student?.avatar.path ||
    reference?.media?.path
  ) {
    avatar = `media/${
      notification?.sender.extraData.avatar ??
      profile?.avatar?.path ??
      student?.avatar?.path ??
      reference?.media?.path
    }?eqmt=${mediaToken}`
  } else {
    avatar = university?.avatar ?? notification?.sender.extraData.avatar ?? null
  }

  const interests =
    profile?.interests ??
    student?.interests ??
    notification?.sender.extraData.interests ??
    null

  const radius = 100 / (2 * Math.PI)
  const diameter = radius * 2

  if (loading) {
    return (
      <mui.Skeleton
        variant="circular"
        width={64}
        height={64}
        className={`rounded-circle`}
      />
    )
  }

  return (
    <div className={Styles.container}>
      {avatar && (
        <img
          className={classNames({ [Styles.border]: !!interests })}
          src={`${process.env.REACT_APP_API_URL}/${avatar}`}
          alt={``}
        />
      )}

      {(profile || student || notification?.sender.role === Role.Student) &&
        !avatar && (
          <img
            className={classNames({ [Styles.border]: !!interests })}
            src={getAvatar(
              profile?.privateAvatar ??
                student?.privateAvatar ??
                (notification?.sender as StudentNotificationSender | undefined)
                  ?.extraData.privateAvatar ??
                1,
            )}
            alt={``}
          />
        )}

      {!profile &&
        !student &&
        notification?.sender.role !== Role.Student &&
        !avatar && (
          <span className={Styles.initials}>
            {nameAbbreviation(
              university?.name ??
                notification?.sender.extraData.name ??
                `${reference?.firstName} ${reference?.lastName}`,
              !!university || notification?.sender.role === Role.University,
            )}
          </span>
        )}

      {interests && (
        <svg className={Styles.ring} viewBox={`0 0 33 33`}>
          {interests.map((interest, i) => (
            <g
              key={`ringColor-${interest}`}
              fill={`none`}
              strokeWidth={1}
              transform={`rotate(${(i / interests.length) * 360} 16.5 16.5)`}
            >
              <path
                d={`M16.5 ${
                  (33 - diameter) / 2
                } a ${radius} ${radius} 0 0 1 0 ${diameter} a ${radius} ${radius} 0 0 1 0 ${
                  diameter * -1
                }`}
                stroke={INTERESTS_COLOR[interest]}
                strokeLinecap={`round`}
                strokeDasharray={`${(1 / interests.length) * 100}, 100`}
              />
            </g>
          ))}
        </svg>
      )}
    </div>
  )
}

export default ProfilePhoto
