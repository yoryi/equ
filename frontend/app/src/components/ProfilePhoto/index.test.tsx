//Testing
import { render } from "@testing-library/react"
import faker from "faker"
import * as React from "react"
//Redux store
import { Provider } from "react-redux"
import configureStore from "redux-mock-store"

//Types
import {
  FollowRequestNotification,
  FollowRequestStatus,
  NotificationType,
  ReduxState,
  Reference,
  ReferenceType,
  Role,
  StudentProfile,
  UniversityProfile,
} from "../../store/types"
//Components
import ProfilePhoto from "./"

type TestStore = Pick<ReduxState, "media">

describe(`ProfilePhoto component`, () => {
  faker.seed(1000)

  const mockStore = configureStore<TestStore>()
  const store = mockStore({
    media: {
      mediaToken: faker.random.alphaNumeric(32),
    },
  })

  const wrapper: React.FC = ({ children }) => (
    <Provider store={store}>{children}</Provider>
  )

  describe(`should render and match the snapshot`, () => {
    let student: StudentProfile
    beforeEach(() => {
      student = {
        id: faker.random.number(10000),
        fullName: faker.name.findName(),
        email: faker.internet.email(),
        schoolName: `${faker.lorem.word()} School`,
        schoolCity: faker.address.city(),
        schoolState: faker.address.stateAbbr(),
        graduation: faker.date.future(1, new Date(1612853271000)).getFullYear(),
        avatar: {
          path: faker.image.imageUrl(),
        },
        interests: [1],
        score: faker.random.number({ min: 75, max: 100 }),
      }
    })

    describe(`student`, () => {
      it(`profile photo set`, () => {
        const { container } = render(<ProfilePhoto student={student} />, {
          wrapper,
        })

        expect(container.firstChild).toMatchSnapshot()
      })

      it(`profile photo not set`, () => {
        student = {
          ...student,
          avatar: {
            path: undefined,
          },
        }

        const { container } = render(<ProfilePhoto student={student} />, {
          wrapper,
        })

        expect(container.firstChild).toMatchSnapshot()
      })
    })

    describe(`university`, () => {
      let university: UniversityProfile
      beforeEach(() => {
        university = {
          id: faker.random.number(10000),
          name: `${faker.lorem.word()} University`,
          city: faker.address.city(),
          state: faker.address.stateAbbr(),
          avatar: faker.image.imageUrl(),
          score: faker.random.number({ min: 75, max: 100 }),
          email: faker.internet.email(),
          website: faker.internet.url(),
        }
      })

      it(`profile photo set`, () => {
        const { container } = render(<ProfilePhoto university={university} />, {
          wrapper,
        })

        expect(container.firstChild).toMatchSnapshot()
      })

      it(`profile photo not set`, () => {
        university = {
          ...university,
          avatar: null,
        }

        const { container } = render(<ProfilePhoto university={university} />, {
          wrapper,
        })

        expect(container.firstChild).toMatchSnapshot()
      })
    })

    describe(`notification`, () => {
      let notification: FollowRequestNotification
      beforeEach(() => {
        notification = {
          id: faker.random.number(10000),
          status: FollowRequestStatus.Pending,
          createdAt: faker.date.past(1, new Date(1612853271000)).toJSON(),
          readAt: null,
          recipient: {
            id: faker.random.number(10000),
            email: faker.internet.email(),
            role: Role.University,
            extraData: null,
          },
          actionType: NotificationType.FollowRequest,
          sender: {
            id: faker.random.number(10000),
            role: Role.Student,
            extraData: {
              studentId: faker.random.number(10000),
              name: faker.name.findName(),
              avatar: faker.image.imageUrl(),
              privateAvatar: faker.random.number({
                min: 1,
                max: 16,
              }),
              interests: [1],
            },
          },
        }
      })

      it(`profile photo is set`, () => {
        const { container } = render(
          <ProfilePhoto notification={notification} />,
          { wrapper },
        )

        expect(container.firstChild).toMatchSnapshot()
      })

      it(`profile photo is not set`, () => {
        notification = {
          ...notification,
          sender: {
            id: faker.random.number(10000),
            role: Role.Student,
            extraData: {
              studentId: faker.random.number(10000),
              name: faker.name.findName(),
              avatar: undefined,
              privateAvatar: faker.random.number({
                min: 1,
                max: 16,
              }),
              interests: [1],
            },
          },
        }

        const { container } = render(
          <ProfilePhoto notification={notification} />,
          { wrapper },
        )

        expect(container.firstChild).toMatchSnapshot()
      })
    })

    describe(`reference`, () => {
      let reference: Reference
      beforeEach(() => {
        reference = {
          id: faker.random.number(10000),
          type: ReferenceType.Academic,
          firstName: faker.name.firstName(),
          lastName: faker.name.lastName(),
          media: {
            id: faker.random.number(10000),
            path: faker.image.imageUrl(),
          },
          organization: faker.company.companyName(),
          position: faker.lorem.word(),
          content: faker.lorem.sentences(3),
          isVisible: false,
          referenceLink: {
            id: faker.random.number(10000),
            type: ReferenceType.Academic,
            activity: null,
            token: faker.random.alphaNumeric(32),
            isActive: false,
            createdAt: null,
          },
        }
      })

      it(`media is set`, () => {
        const { container } = render(<ProfilePhoto reference={reference} />, {
          wrapper,
        })

        expect(container.firstChild).toMatchSnapshot()
      })

      it(`media is not set`, () => {
        reference = {
          ...reference,
          media: null,
        }

        const { container } = render(<ProfilePhoto reference={reference} />, {
          wrapper,
        })

        expect(container.firstChild).toMatchSnapshot()
      })
    })

    it(`loading`, () => {
      const { container } = render(<ProfilePhoto loading />, { wrapper })

      expect(container.firstChild).toMatchSnapshot()
    })
  })
})
