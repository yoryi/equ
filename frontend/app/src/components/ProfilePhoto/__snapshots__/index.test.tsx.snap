// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfilePhoto component should render and match the snapshot loading 1`] = `
<span
  class="MuiSkeleton-root MuiSkeleton-circular MuiSkeleton-pulse rounded-circle css-fomgco-MuiSkeleton-root"
  style="width: 64px; height: 64px;"
/>
`;

exports[`ProfilePhoto component should render and match the snapshot notification profile photo is not set 1`] = `
<div
  class="container"
>
  <img
    alt=""
    class="border"
    src="13.svg"
  />
  <svg
    class="ring"
    viewBox="0 0 33 33"
  >
    <g
      fill="none"
      stroke-width="1"
      transform="rotate(0 16.5 16.5)"
    >
      <path
        d="M16.5 0.5845056908104667 a 15.915494309189533 15.915494309189533 0 0 1 0 31.830988618379067 a 15.915494309189533 15.915494309189533 0 0 1 0 -31.830988618379067"
        stroke="#005DCC"
        stroke-dasharray="100, 100"
        stroke-linecap="round"
      />
    </g>
  </svg>
</div>
`;

exports[`ProfilePhoto component should render and match the snapshot notification profile photo is set 1`] = `
<div
  class="container"
>
  <img
    alt=""
    class="border"
    src="https://mock.equedi.com/media/http://lorempixel.com/640/480?eqmt=n74ly0hcvu701ce68uuk7rqwe260qw2t"
  />
  <svg
    class="ring"
    viewBox="0 0 33 33"
  >
    <g
      fill="none"
      stroke-width="1"
      transform="rotate(0 16.5 16.5)"
    >
      <path
        d="M16.5 0.5845056908104667 a 15.915494309189533 15.915494309189533 0 0 1 0 31.830988618379067 a 15.915494309189533 15.915494309189533 0 0 1 0 -31.830988618379067"
        stroke="#005DCC"
        stroke-dasharray="100, 100"
        stroke-linecap="round"
      />
    </g>
  </svg>
</div>
`;

exports[`ProfilePhoto component should render and match the snapshot reference media is not set 1`] = `
<div
  class="container"
>
  <span
    class="initials"
  >
    LB
  </span>
</div>
`;

exports[`ProfilePhoto component should render and match the snapshot reference media is set 1`] = `
<div
  class="container"
>
  <img
    alt=""
    class=""
    src="https://mock.equedi.com/media/http://lorempixel.com/640/480?eqmt=n74ly0hcvu701ce68uuk7rqwe260qw2t"
  />
</div>
`;

exports[`ProfilePhoto component should render and match the snapshot student profile photo not set 1`] = `
<div
  class="container"
>
  <img
    alt=""
    class="border"
    src="01.svg"
  />
  <svg
    class="ring"
    viewBox="0 0 33 33"
  >
    <g
      fill="none"
      stroke-width="1"
      transform="rotate(0 16.5 16.5)"
    >
      <path
        d="M16.5 0.5845056908104667 a 15.915494309189533 15.915494309189533 0 0 1 0 31.830988618379067 a 15.915494309189533 15.915494309189533 0 0 1 0 -31.830988618379067"
        stroke="#005DCC"
        stroke-dasharray="100, 100"
        stroke-linecap="round"
      />
    </g>
  </svg>
</div>
`;

exports[`ProfilePhoto component should render and match the snapshot student profile photo set 1`] = `
<div
  class="container"
>
  <img
    alt=""
    class="border"
    src="https://mock.equedi.com/media/http://lorempixel.com/640/480?eqmt=n74ly0hcvu701ce68uuk7rqwe260qw2t"
  />
  <svg
    class="ring"
    viewBox="0 0 33 33"
  >
    <g
      fill="none"
      stroke-width="1"
      transform="rotate(0 16.5 16.5)"
    >
      <path
        d="M16.5 0.5845056908104667 a 15.915494309189533 15.915494309189533 0 0 1 0 31.830988618379067 a 15.915494309189533 15.915494309189533 0 0 1 0 -31.830988618379067"
        stroke="#005DCC"
        stroke-dasharray="100, 100"
        stroke-linecap="round"
      />
    </g>
  </svg>
</div>
`;

exports[`ProfilePhoto component should render and match the snapshot university profile photo not set 1`] = `
<div
  class="container"
>
  <span
    class="initials"
  >
    UU
  </span>
</div>
`;

exports[`ProfilePhoto component should render and match the snapshot university profile photo set 1`] = `
<div
  class="container"
>
  <img
    alt=""
    class=""
    src="https://mock.equedi.com/http://lorempixel.com/640/480"
  />
</div>
`;
